staging/src/k8s.io/apiserver/pkg/cel/library/cidr.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package library
+
+import (
+	"fmt"
+	"net/netip"
+	"strings"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+
+	apiservercel "k8s.io/apiserver/pkg/cel"
+)
+
+// CIDR provides a CEL function library extension of CIDR notation parsing functions.
+//
+// cidr
+//
+// Converts a string in CIDR notation to a network address representation or results in an error if the string is not a valid CIDR notation.
+// The CIDR must be an IPv4 or IPv6 subnet address with a mask.
+// Leading zeros in IPv4 address octets are not allowed.
+// IPv4-mapped IPv6 addresses (e.g. ::ffff:1.2.3.4/24) are not allowed.
+//
+//	cidr(<string>) <CIDR>
+//
+// Examples:
+//
+//	cidr('192.168.0.0/16') // returns an IPv4 address with a CIDR mask
+//	cidr('::1/128') // returns an IPv6 address with a CIDR mask
+//	cidr('192.168.0.0/33') // error
```

Comment:
> What does an error mean for the expression evaluator? Is it coverted to "false" or something else?

Any error in parsing is returned to the user as an error and fails the validation, rejecting the request. This means you wouldn't be able to conditionally have logic such as `isCIDR(self) ? cidr(self).bits() == 32 : isIP(self)` to check that if a CIDR is passed then it would be a single address. Not sure you'd want to use that, but this would error if I replaced `isCIDR` with just `cidr`.

> I ask because isCIDR(s) seems OK, but maybe not needed if cidr(s) would produce an equivalent result. It seems like every time we would check isCIDR we also need the result AS a CIDR, so we would parse once to satisfy is and again to convert.

I have followed the patterns set out by the other libraries, such as the URL library. In all the examples I've seen, the parse happens many times. For example, the [CEL playground](https://playcel.undistro.io/?content=H4sIAAAAAAAAA31RW0vDMBT%2BK4cOdgHX6JsUhoIMBPcwFd%2FycpqettE0iUk6Nsf%2Bu%2BnWTfewQR6Sj%2B%2BWc7ZJgQGTLNlyDcATk3%2BSCDzJYA90UO2o7IB4C8H6jLHWk8ssev9Ia2ysolSYJru%2FZRZD%2FfDdktvMVqgGpcOqIR3tOq8d1%2FEkNwmtrSPvpdExlzGYH0w8mBJe2pycphBfH28LeJovQMncodtAqDGA9IArlApzRSB1BAmswk3lTKuLlOvod6z5dfJKpWGFEZ7Fr5AjLSj%2BQepqilYyQYoN%2FrjT1qlpn8k119LHIuPDXNJuFhPgejiESDtD04rCu6ipofEEZjMY7WuMLpOfjQ899XyMo6sSjaeEf7LLmqVxx5hr1nMv0FKxjBvs2fttXha8dmseT1Ivf%2Fo%2Bd3G7u1%2Bd0weKUQIAAA%3D%3D) example parses the URL 7 times.

As far as I understand there is no way to declare a variable or or optimise this so that we only parse once for complex expressions involving these.